quickwit/quickwit-control-plane/src/ingest/ingest_controller.rs

```diff
@@ -706,9 +780,189 @@ impl IngestController {
+            .init_shards(&open_shards_response.subresponses, progress)
+            .await;
+
+        for init_shard_success in init_shards_response.successes {
+            let shard = init_shard_success.shard().clone();
+            let index_uid = shard.index_uid().clone();
+            let source_id = shard.source_id.clone();
+            model.insert_shards(&index_uid, &source_id, vec![shard.clone()]);
+
+            let source_uid = SourceUid {
+                index_uid,
+                source_id,
+            };
+            // We temporarily disable the ability the scale down the number of shards for the source
+            // to avoid closing the shards we just opened.
+            model.drain_scaling_permits(&source_uid, ScalingMode::Down);
+        }
+        for init_shard_failure in init_shards_response.failures {
+            let shard_id = init_shard_failure.shard_id();
+            shards_to_close.remove(shard_id);
+        }
+        let mut per_leader_shards_to_close: HashMap<String, Vec<ShardPKey>> = HashMap::new();
+
+        for (leader_id, shard_pkey) in shards_to_close.into_values() {
+            per_leader_shards_to_close
+                .entry(leader_id)
+                .or_default()
+                .push(shard_pkey);
+        }
+        let mut close_shards_futures = FuturesUnordered::new();
+
+        for (leader_id, shard_pkeys) in per_leader_shards_to_close {
+            let Some(mut ingester) = self.ingester_pool.get(&leader_id) else {
+                warn!("failed to close shards: ingester `{leader_id}` is unavailable");
+                continue;
+            };
+            let shard_ids = shard_pkeys
+                .iter()
+                .map(|shard_pkey| ShardIds {
+                    index_uid: shard_pkey.index_uid.clone(),
+                    source_id: shard_pkey.source_id.clone(),
+                    shard_ids: shard_pkey.shard_id.clone().into_iter().collect(),
+                })
+                .collect();
+            #[allow(deprecated)]
+            let shards_to_close_request = CloseShardsRequest {
+                shard_pkeys,
+                shard_ids,
+            };
+            let close_shards_future = async move {
+                tokio::time::timeout(
+                    Duration::from_secs(3),
+                    ingester.close_shards(shards_to_close_request),
+                )
+                .await
+            };
+            close_shards_futures.push(close_shards_future);
+        }
+        let mailbox = mailbox.clone();
+        let close_shards_future = async move {
+            // We wait for a few seconds before closing the shards to give the ingesters some time
+            // to learn about the ones we just opened via gossip.
+            tokio::time::sleep(Duration::from_secs(10)).await;
+
+            let mut closed_shards = Vec::new();
+
+            while let Some(close_shards_result) = close_shards_futures.next().await {
```

Comment:
You have an individual timeout on  the different close shard request, but no differentiated logging.

We should either have specific loggging information, or make the timeout global for the set of futures.

The benefit of the global timeout, is that you can move it closed to what's happening in the task, and help proofreading.

A reader wants to be able to answer questions like, "is it possible to hold the rebalance permit forever? Is it capped in time?"